src/+matlode/+rk/+erk/ERK.m

```diff
@@ -1,99 +1,304 @@
             
-            opts = obj.matlodeSets(varargin{:});
+            %ERK specific options
+            
+            opts = matlodeSets@matlode.rk.RungeKutta(obj, p, varargin{:});
+        end
+        
+        function [t, y, stats] = timeLoop(obj, f, tspan, y0, opts, varargin)
             
             numVars = length(y0);
-            t = tspan(1);
-            k = zeros(numVars, obj.stage);
-            sol.y = y0;
+            multiTspan = length(tspan) > 2;
+            isDense = ~isempty(opts.Dense);
+            overstep = false;
+            startupErr = true;
+            
+            %Structure access optimizations
+            errNormFunc = opts.ErrNorm;
+            stepController = opts.StepSizeController;
+            newStepFunc = @(prevAccept, tCur, tspan, hHist, err, q, nSteps, nFailed) opts.StepSizeController.newStepSize(prevAccept, tCur, tspan, hHist, err, q, nSteps, nFailed);
+            
+            
+            k = zeros(numVars, obj.Stage);
+            y = [];
+            t = [];
             
-            sol.stats.nsteps = 0;
-            sol.stats.nfailed = 0;
-            sol.stats.nfevals = 0;
+            if multiTspan
+                y = zeros(numVars, length(tspan));
+                t = zeros(1, length(tspan));
+            else
+                y = zeros(numVars, opts.ChunkSize);
+                t = zeros(1, opts.ChunkSize);
+            end
+            
+            %inital values
+            t(1,1) = tspan(1);
+            tCur = tspan(1);
+            tNext = tCur;
+            ti = 2;
+            hmax = opts.MaxStep;
+            
+            y(:, 1) = y0;
+            yNext = y0;
+            
+            %temporary stats
+            nSteps = 1;
+            nFailed = 0;
+            nFevals = 1;
+            nSmallSteps = 0;
+            
+            tdir = sign(tspan(end) - tspan(1));
+            
+            q = min(obj.Order, obj.EmbeddedOrder);
+            
+            %%%%
+            % Start of method specific intializing code
+            %%%%
             
             fsalStart = uint32(obj.FSAL) + 1;
-            fevalIterCounts = obj.stage - fsalStart + 1;
+            fevalIterCounts = double(obj.Stage - fsalStart + 1);
             
+            %%%%
+            % End of method specific intializing code
+            %%%%
             
-            [h, k(:,1)] = opts.StepSizeController.startingStep(f, tspan, y0, obj.order, opts);
+            %First step
+            %allocates memory for first step
+            [hHist, k(:,end)] = stepController.startingStep(f, tspan, y0, obj.Order, errNormFunc, opts.InitialStep);
+            hN = hHist(1);
+            err = zeros(1, length(hHist));
             
+            accept = true;
             
-            while t < tspan(end)
+            %Time Loop
+            while ti <= length(tspan)
+                %Cycle history
+                err = circshift(err, 1, 2);
```

Comment:
cyclic indexing would be using our own matlab setup, circshift is a C implemented function. It could be just moving a reference pointer instead which is pretty fast

Reply:
It's also a small vector, so it should just be fast in general, even if it does copies.